src/openloco/viewportmgr.cpp

```diff
@@ -70,16 +168,243 @@ namespace openloco::ui::viewportmgr
+                continue;
+
+            if (!viewport->intersects(rect))
+                continue;
+
+            // Get intersection
+            auto left = std::max(rect.left, viewport->view_x);
+            auto right = std::min<int16_t>(rect.right, viewport->view_x + viewport->view_width);
+            auto top = std::max(rect.top, viewport->view_y);
+            auto bottom = std::min<int16_t>(rect.bottom, viewport->view_y + viewport->view_height);
+
+            // offset rect by (negative) viewport origin
+            left -= viewport->view_x;
+            right -= viewport->view_x;
+            top -= viewport->view_y;
+            bottom -= viewport->view_y;
+
+            // apply zoom
+            left = left >> viewport->zoom;
+            right = right >> viewport->zoom;
+            top = top >> viewport->zoom;
+            bottom = bottom >> viewport->zoom;
+
+            // offset calculated area by viewport offset
+            left += viewport->x;
+            right += viewport->x;
+            top += viewport->y;
+            bottom += viewport->y;
+
+            gfx::set_dirty_blocks(left, top, right, bottom);
+        }
+    }
+
+    // 0x004CBA2D
+    void invalidate(station* station)
+    {
+        for (auto& i : _viewports)
+        {
+            auto viewport = i.get();
+
+            if (viewport == nullptr)
+                break;
+
+            ViewportRect rect;
+            rect.left = station->var_08[viewport->zoom];
+            rect.top = station->var_18[viewport->zoom];
+            rect.right = station->var_10[viewport->zoom] + 1;
+            rect.bottom = station->var_20[viewport->zoom] + 1;
+
+            rect.left <<= viewport->zoom;
+            rect.top <<= viewport->zoom;
+            rect.right <<= viewport->zoom;
+            rect.bottom <<= viewport->zoom;
+
+            if (!viewport->intersects(rect))
+                continue;
+
+            // Get intersection
+            auto left = std::max(rect.left, viewport->view_x);
+            auto right = std::min<int16_t>(rect.right, viewport->view_x + viewport->view_width);
+            auto top = std::max(rect.top, viewport->view_y);
+            auto bottom = std::min<int16_t>(rect.bottom, viewport->view_y + viewport->view_height);
+
+            // offset rect by (negative) viewport origin
+            left -= viewport->view_x;
+            right -= viewport->view_x;
+            top -= viewport->view_y;
+            bottom -= viewport->view_y;
+
+            // apply zoom
+            left = left >> viewport->zoom;
+            right = right >> viewport->zoom;
+            top = top >> viewport->zoom;
+            bottom = bottom >> viewport->zoom;
+
+            // offset calculated area by viewport offset
+            left += viewport->x;
+            right += viewport->x;
+            top += viewport->y;
+            bottom += viewport->y;
+
+            gfx::set_dirty_blocks(left, top, right, bottom);
```

Comment:
Could also use the ViewportRect as argument, clarifies that the coordinates are screen space.

Reply:
I'd like to have ViewportRect separate from ScreenRect. Viewports are in "rendered world" space, while the screen is like... a layer on top of that. If that makes sense.